File: 3 Longest Substring Without Repeating Characters/Python/solution.py
from typing import Self, Dict


class Solution:
    def lengthOfLongestSubstring(self: Self, s: str) -> int:
        max_len: int = 0
        start_idx: int = 0
        past_chars: Dict[str, int] = {}

        for end_idx in range(len(s)):
            curr_char: str = s[end_idx]

            if curr_char in past_chars and past_chars[curr_char] >= start_idx:
                start_idx = past_chars[curr_char] + 1

            past_chars[curr_char] = end_idx
            max_len = max(max_len, end_idx - start_idx + 1)
            
        return max_len

File: 3 Longest Substring Without Repeating Characters/Java/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxLen = 0;
        int startIdx = 0;
        Map<Character, Integer> pastChars = new HashMap<>();

        for (int endIdx = 0; endIdx < s.length(); endIdx++) {
            char currentChar = s.charAt(endIdx);

            if (pastChars.containsKey(currentChar) && pastChars.get(currentChar) >= startIdx) {
                startIdx = pastChars.get(currentChar) + 1;
            }

            pastChars.put(currentChar, endIdx);
            maxLen = Math.max(maxLen, endIdx - startIdx + 1);
        }

        return maxLen;
    }
}
